File: Problems/python/boj_단지번호붙이기.py
#https://www.acmicpc.net/problem/2667
#BFS

from collections import deque

n = int(input())
mat = [list(map(int, input())) for _ in range(n)]

moves = ((1,0), (-1,0), (0,1), (0,-1))
def bfs(start, m):
    q = deque([start])
    x, y = start
    m[x][y] = -1
    count = 1

    while q:
        x, y = q.popleft()
        for i, j in moves:
            nx = x + i
            ny = y + j
            if not 0<=nx<n or not 0<=ny<n:
                continue
            if m[nx][ny] == 1:
                m[nx][ny] = -1
                q.append((nx, ny))
                count += 1
    return count
        
result = []
for i in range(n):
    for j in range(n):
        if mat[i][j] == 1:
            result.append(bfs((i,j), mat))

result.sort()
print(len(result))
for i in result:
    print(i)
            
        

File: Problems/cpp/boj_단지번호붙이기.cpp
//https://www.acmicpc.net/problem/2667
//DFS, BFS


#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <algorithm>

using namespace std;

int n, cnt=0;
vector<int> answer;
vector<string> mat;
vector<vector<bool>> visit;
int dx[] = {0, 0, -1, 1};
int dy[] = {1, -1, 0, 0};

int bfs(int i, int j){
    int cnt = 1;
    deque<pair<int, int>> dq;
    dq.push_back(make_pair(i, j));
    visit[i][j] = true;

    while (!dq.empty()){
        int x = dq.front().first;
        int y = dq.front().second;
        dq.pop_front();

        for (int i=0; i<4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (0>nx || 0>ny || nx>=n || ny>=n) continue;

            if (mat[nx][ny]=='1' && !visit[nx][ny]){
                cnt++;
                visit[nx][ny] = true;
                dq.push_back(make_pair(nx, ny));
            }
        }
    }

    return cnt;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> n;

    mat.resize(n);
    visit.resize(n, vector<bool>(n, false));
    for (int i=0; i<n; i++)
        cin >> mat[i];

    for (int i=0; i<n; i++){
        for (int j=0; j<n; j++){
            if (mat[i][j] == '1' && !visit[i][j]){
                answer.push_back(bfs(i, j));
                cnt++;
            }
        }
    }

    sort(answer.begin(), answer.end());

    cout << cnt << '\n';
    for (const int& a : answer)
        cout << a << '\n';
}